File: PyCommander/telemetry/packets/clock_task.py
from ..topic_packets import ClockPacket
from ..base_packet import BasePacket, get_day_millis
from ..command_schema import Topic, ClockCMD

from ..serial_comms import SerialComms, DEFAULT_PORT

from .packet_task import PacketTask

class ClockTask(PacketTask):

    @staticmethod
    def route_packet(packet: ClockPacket) -> ClockCMD:

        if (packet.topic != Topic.SYSTEM_STATUS.value):
            print(f"Routing - BasePacket Topic {packet.topic} not {Topic.SYSTEM_STATUS.value}")
            
        command = ClockCMD(packet.command)

        print(command)

        match command:
            case ClockCMD.STATUS:
                print("REQUEST ACK")
                pass

            case ClockCMD.DAY_SYNC:
                print("ACK")
                pass

            case ClockCMD.JUMP_CLOCK_TELEM:
                print("JUMP_CLOCK_TELEM")
                pass

            case ClockCMD.JUMP_CLOCK_GPS:
                print("JUMP_CLOCK_GPS")
                pass

            case _:
                print("Did not route packet")
                return 
            
        return command
    
    def send_jump_clock(offset_ms: int = 0):
        
        millis, date_time = get_day_millis()
        message: list[int] = []

        # # If adding millistamp
        # message.append( (millis >> 24) & 0xFF)
        # message.append( (millis >> 16) & 0xFF)
        # message.append( (millis >> 8) & 0xFF)
        # message.append( millis & 0xFF )

        message.append(date_time.day) # 1-31
        message.append(date_time.month) # 1-12
        message.append( (date_time.year >> 8) & 0xFF) # 1-9999
        message.append(date_time.year & 0xFF) # 1-9999

        packet = ClockPacket()
        packet.configure(ClockCMD.JUMP_CLOCK_TELEM, bytearray(message))
        packet.packetize()

        DEFAULT_PORT.emit_packet(packet)

        

File: MicrocontrollerCode/src/common/telemetry_tasks/ClockTask.cpp
#include "ClockTask.h"
#include "SystemStatusTask.h"
#include "../comms/serial_comms.h"

using namespace std;

namespace Cesium {

size_t ClockTask::routed_packets = 0;

ClockCMD ClockTask::route_packet(BasePacket &packet)
{
    ClockCMD command = (ClockCMD)packet.get_command();
    switch(command) {
    case ClockCMD::STATUS:
        DEBUGLN("STATUS");
        SystemStatusTask::send_not_implemented("CLOCK::STATUS");
        break;

    case ClockCMD::DAY_SYNC: // No further action
        DEBUGLN("DAY_SYNC");
        SystemStatusTask::send_not_implemented("CLOCK::DAY_SYNC");
        break;

    case ClockCMD::JUMP_CLOCK_TELEM: // No further action
        DEBUGLN("JUMP_CLOCK_TELEM");
        SystemStatusTask::send_ack("CLOCK::JUMP_CLOCK_TELEM");
        jump_clock_telem(packet);
        break;

    case ClockCMD::JUMP_CLOCK_GPS: // TODO
        DEBUGLN("JUMP_CLOCK_GPS");
        SystemStatusTask::send_not_implemented("CLOCK::JUMP_CLOCK_GPS");
        break;

    default:
        DEBUGLN("Did not finish routing packet");
        SystemStatusTask::send_nack("BAD COMMAND");
        return ClockCMD(-1);
    }

    return command;
}

bool ClockTask::jump_clock_telem(BasePacket &packet)
{
    vector<uint8_t> data = packet.get_data();

    // uint8_t, uint8_t, uint16_t
    // day      month    year
    size_t data_size = data.size();

    // Verifies data length
    if (data_size != 4) {
        String err = "Data size '" + String(data_size) + "' when 4 expected";
        DEBUGLN(err);
        send_jump_clock_response(err);
        return false;
    }

    uint8_t day = data[0];
    uint8_t month = data[1];
    uint16_t year = data[2];

    // Set clock of ESP32
    RETURN_FALSE_IF_FALSE(Clock::jump_clock(packet.get_millistamp(), day, month, year));

    send_jump_clock_response("success");
    
    return true;
}

void ClockTask::send_jump_clock_response(String message)
{
    vector<uint8_t> data(message.c_str(), message.c_str() + message.length());

    BasePacket packet;
    packet.configure((int)Topic::CLOCK, (int)ClockCMD::JUMP_CLOCK_TELEM, data);
    packet.packetize();

    SerialComms::emit_packet(packet);
}

}